syntax = "proto3";

package orders;

service OrderService {
  rpc SubmitOrder (OrderRequest) returns (OrderResponse);
  rpc ModifyOrder (ModifyOrderRequest) returns (OrderResponse);
  rpc CancelOrder (CancelOrderRequest) returns (OrderResponse);
  rpc GetLob (LobRequest) returns (LobResponse);
}

message LobRequest {
  string ticker = 1;
}

message PriceLevel {
  double price = 1;
  int32 total_qty = 2;
}

message LobResponse {
  string ticker = 1;
  repeated PriceLevel bids = 2;
  repeated PriceLevel asks = 3;
}

message OrderRequest {
  string tikcer = 1;
  string type = 2; 
  double price = 3;
  int32 quantity = 4;
}

message ModifyOrderRequest {
  string order_id = 1;
  double new_price = 2;
  int32 new_quantity = 3;
}

message CancelOrderRequest {
  string order_id = 1;
}
message Trade {
  string unique_id = 1;
  int64 timestamp = 2;
  double price = 3;
  int32 qty = 4; 
  string bid_order_id = 5;
  string ask_order_id = 6;
}

message OrderResponse {
  string status = 1;
  string order_id = 2;
  repeated Trade trades = 3;
  
}